tests/foreman/api/test_remoteexecution.py

```diff
@@ -0,0 +1,121 @@
+from robottelo.vm_capsule import CapsuleVirtualMachine
+
+
+@tier4
+def test_positive_run_capsule_upgrade_playbook():
+    """Run Capsule Upgrade playbook against an External Capsule
+
+    :id: 9ec6903d-2bb7-46a5-8002-afc74f06d83b
+
+    :steps:
+        1. Create a Capsule VM, add REX key.
+        2. Run the Capsule Upgrade Playbook.
+
+    :expectedresults: Capsule is upgraded successfully
+
+    :CaseImportance: Medium
+    """
+    with CapsuleVirtualMachine() as capsule_vm:
+        add_remote_execution_ssh_key(capsule_vm.ip_addr)
+        job = entities.JobInvocation().run(
+            synchronous=False,
+            data={
+                'job_template_id': 170,
+                'inputs': {
+                    'target_version': "6.8.z",
+                    'whitelist_options': "repositories-validate,repositories-setup",
+                },
+                'targeting_type': "static_query",
+                'search_query': f"name = {capsule_vm.hostname}",
+            },
+        )
+        wait_for_tasks(f"resource_type = JobInvocation and resource_id = {job['id']}")
+        result = entities.JobInvocation(id=job['id']).read()
+        assert result.succeeded == 1
+
+        result = capsule_vm.run('foreman-maintain health check')
+        assert result.return_code == 0
+        for line in result.stdout:
+            assert 'FAIL' not in line
+
+        result = entities.SmartProxy(
+            id=entities.SmartProxy(name=capsule_vm.hostname).search()[0].id
+        ).refresh()
+        feature_list = [feat['name'] for feat in result['features']]
+        assert set(['SSH', 'TFTP', 'HTTPBoot', 'Dynflow', 'Pulp Node', 'Logs']).issubset(
+            feature_list
+        )
+
+
+@destructive
+def test_negative_run_capsule_upgrade_playbook_on_satellite(default_org):
+    """Run Capsule Upgrade playbook against the Satellite itself
+
+    :id: 99462a11-5133-415d-ba64-4354da539a34
+
+    :steps:
+        1. Add REX key to the Satellite server.
+        2. Run the Capsule Upgrade Playbook.
+        3. Check the job output for proper failure reason.
+
+    :expectedresults: Should fail
+
+    :CaseImportance: Medium
+    """
+    sat = entities.Host().search(query={'search': f'name={settings.server.hostname}'})[0].read()
+    template_id = (
+        entities.JobTemplate().search(query={'search': 'name="Capsule Upgrade Playbook"'})[0].id
+    )
+
+    add_remote_execution_ssh_key(sat.name)
+    try:
+        entities.JobInvocation().run(
+            data={
+                'job_template_id': template_id,
+                'inputs': {
+                    'target_version': "6.8.z",
+                    'whitelist_options': "repositories-validate,repositories-setup",
+                },
+                'targeting_type': "static_query",
+                'search_query': f"name = {sat.name}",
+            }
+        )
+    except TaskFailedError as err:
+        assert 'A sub task failed' in err.args[0]
+    finally:
```

Comment:
Do we need this `finally`?
I think you could use [`pytest.raises`](https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions) to catch and assert on exception, and let rest of the code run unconditionally. `pytest.raises` will also tell you if function **did not** raise exception.

Reply:
Yes, it's much nicer with `pytest.raises`, thank you for the tip! Update pushed.